### PR TITLE
Add preference to hide Android system cursor

### LINKED ISSUE #28: [Feature Request] Option to hide local Android cursor
## Summary
 * Currently, when an external mouse is connected, the local Android system cursor completely covers the remote cursor

## Proposed Solution
 * An option to hide the local cursor would therefore be useful!

## Alternative Solutions
 * Not making this a setting and always hiding the local cursor
   * This would make it harder to debug input issues.

## Version Information
AVNC 1.4.0

### PR DESCRIPTION
**Pull request by personalizedrefrigerator**
## Summary
 * Adds a new preference that hides the default system cursor (on by default)
 * Requires Android N or above &mdash; includes API level checks (only tested on a more-recent Android device)
 * Should fix #28

## TODO
 * [ ] Check that the API level checks prevent things from breaking if `Build.VERSION.SDK_INT ≤ Build.VERSION_CODES.N`

### CODE REVIEW

File: app/src/main/java/com/gaurav/avnc/ui/prefs/PrefsActivity.kt [hunk 1]
```diff
@@ -92,7 +92,21 @@ class PrefsActivity : AppCompatActivity(), PreferenceFragmentCompat.OnPreference
         }
     }
 
-    @Keep class Input : PrefFragment(R.xml.pref_input)
+    @Keep class Input : PrefFragment(R.xml.pref_input) {
+        override fun onCreate(savedInstanceState: Bundle?) {
+            super.onCreate(savedInstanceState)
+
+            val canChangePtrIcon = Build.VERSION.SDK_INT >= Build.VERSION_CODES.N
```
Comment (gujjwal00) [inline app/src/main/java/com/gaurav/avnc/ui/prefs/PrefsActivity.kt:99]:
```suggestion
            val canChangePtrIcon = Build.VERSION.SDK_INT >= 24
```

This is just my personal preference, so if you want, you can ignore this suggestion.

File: app/src/main/java/com/gaurav/avnc/ui/vnc/FrameView.kt [hunk 2]
```diff
@@ -77,6 +84,13 @@ class FrameView(context: Context?, attrs: AttributeSet? = null) : GLSurfaceView(
         frameState.setViewportSize(w.toFloat(), h.toFloat())
     }
 
+    @RequiresApi(Build.VERSION_CODES.N)
+    override fun onResolvePointerIcon(e: MotionEvent, pointerIndex: Int): PointerIcon {
+        if (!prefs.input.hideLocalCursor)
+            return PointerIcon.getSystemIcon(context, PointerIcon.TYPE_DEFAULT)
+
+        return PointerIcon.getSystemIcon(context, PointerIcon.TYPE_NULL)
+    }
```
Comment (gujjwal00) [inline app/src/main/java/com/gaurav/avnc/ui/vnc/FrameView.kt:93]: Instead of overriding `onResolvePointerIcon()`, we should use [setPointerIcon()](https://developer.android.com/reference/android/view/View#setPointerIcon(android.view.PointerIcon)) to hide the pointer. We can update the pointer at the end of [initialize()](https://github.com/gujjwal00/avnc/blob/0188996bc0b126d4dcb727821b6a6fa1dd7ede62/app/src/main/java/com/gaurav/avnc/ui/vnc/FrameView.kt#L73), like this:
```kotlin
        if (Build.VERSION.SDK_INT >= 24 && viewModel.pref.input.hideLocalCursor)
            pointerIcon = PointerIcon.getSystemIcon(context, PointerIcon.TYPE_NULL)
```

Overriding `onResolvePointerIcon()` like this has some drawbacks:
- We are doing a map-lookup via `prefs.input.hideLocalCursor` for every event, even though the lookup result won't  change during the session.
- Its more verbose compared to `setPointerIcon()`, which suits our requirement perfectly.

File: app/src/main/java/com/gaurav/avnc/util/AppPreferences.kt [hunk 3]
```diff
@@ -54,6 +54,7 @@ class AppPreferences(context: Context) {
         val vkShowAll; get() = prefs.getBoolean("vk_show_all", false)
 
         val mousePassthrough; get() = prefs.getBoolean("mouse_passthrough", true)
+        val hideLocalCursor; get() = prefs.getBoolean("hide_local_cursor", true)
```
Comment (gujjwal00) [inline app/src/main/java/com/gaurav/avnc/util/AppPreferences.kt:57]:
```suggestion
        val hideLocalCursor; get() = prefs.getBoolean("hide_local_cursor", false)
```

Please change the default value for this preference to `false`. 
- Existing users (like me), who use the 'Hide Remote Cursor' server option, suddenly won't *any* cursor at all, if this pref is on by default.
- This behavior of hiding cursor is unexpected from user POV, and it interacts badly with the toolbar (try opening the toolbar and using it). Hence it should be opt-in.
- It breaks the experience of users who disable 'Mouse Passthrough' (probably not many, but still).

File: app/src/main/res/xml/pref_input.xml [hunk 4]
```diff
@@ -80,6 +80,13 @@
             app:summaryOff="@string/pref_mouse_passthrough_summary_off"
             app:summaryOn="@string/pref_mouse_passthrough_summary_on"
             app:title="@string/pref_mouse_passthrough" />
+
+        <SwitchPreference
+            app:defaultValue="true"
```
Comment (gujjwal00) [inline app/src/main/res/xml/pref_input.xml:85]:
```suggestion
            app:defaultValue="false"
```

File: app/src/main/res/xml/pref_input.xml [hunk 5]
```diff
@@ -80,6 +80,13 @@
...
+            app:key="hide_local_cursor"
+            app:summaryOff="@string/pref_hide_local_cursor_summary_off"
+            app:summaryOn="@string/pref_hide_local_cursor_summary_on"
```
Comment (gujjwal00) [inline app/src/main/res/xml/pref_input.xml:88]: Summary is unnecessary here, as preference title is clear enough IMO. 
We can even use `Hide Local (Android) Cursor` as title to make it more clear if you want.

**Comment by gujjwal00** [comment]
BTW, thank you for working on this.

**Comment by personalizedrefrigerator** [comment]
Thank you for the review! I think I've made the requested changes.

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -57,6 +57,7 @@
     <string name="msg_server_profile_added">Server added.</string>
     <string name="msg_server_profile_deleted">Server deleted.</string>
     <string name="msg_pip_not_supported">Picture-in-Picture is not supported on this device</string>
+    <string name="msg_change_ptr_not_supported">Pointer styling not supported on this device</string>
```
Comment (gujjwal00) [inline app/src/main/res/values/strings.xml:60]:
```suggestion
    <string name="msg_ptr_hiding_not_supported">Hiding cursor is not supported on this device</string>
```

This message should be aligned with the target preference. 'Pointer styling' & 'Cursor hiding' seems unrelated to me.

File: app/src/main/java/com/gaurav/avnc/ui/prefs/PrefsActivity.kt [hunk 7]
```diff
@@ -92,7 +92,21 @@ class PrefsActivity : AppCompatActivity(), PreferenceFragmentCompat.OnPreference
...
+    @Keep class Input : PrefFragment(R.xml.pref_input) {
+        override fun onCreate(savedInstanceState: Bundle?) {
+            super.onCreate(savedInstanceState)
+
+            val canChangePtrIcon = Build.VERSION.SDK_INT >= 24
+
+            if (!canChangePtrIcon) {
+                findPreference<SwitchPreference>("hide_local_cursor")!!.apply {
+                    isEnabled = false
+                    summary = getString(R.string.msg_change_ptr_not_supported)
```
Comment (gujjwal00) [inline app/src/main/java/com/gaurav/avnc/ui/prefs/PrefsActivity.kt:104]:
```suggestion
                    summary = getString(R.string.msg_ptr_hiding_not_supported)
```
See below.

**Comment by gujjwal00** [comment]
Just a minor nit.

File: app/src/main/java/com/gaurav/avnc/ui/vnc/FrameView.kt [hunk 8]
```diff
@@ -11,12 +11,17 @@ package com.gaurav.avnc.ui.vnc
...
 import android.content.Context
 import android.opengl.GLSurfaceView
+import android.os.Build
 import android.util.AttributeSet
 import android.view.KeyEvent
 import android.view.MotionEvent
+import android.view.PointerIcon
 import android.view.inputmethod.BaseInputConnection
 import android.view.inputmethod.EditorInfo
+import androidx.annotation.RequiresApi
```
Comment (gujjwal00) [inline app/src/main/java/com/gaurav/avnc/ui/vnc/FrameView.kt:21]:
```suggestion
```

File: app/src/main/java/com/gaurav/avnc/ui/vnc/FrameView.kt [hunk 9]
```diff
@@ -11,12 +11,17 @@ package com.gaurav.avnc.ui.vnc
...
 import com.gaurav.avnc.ui.vnc.gl.Renderer
+import com.gaurav.avnc.util.AppPreferences
+import com.gaurav.avnc.viewmodel.PrefsViewModel
```
Comment (gujjwal00) [inline app/src/main/java/com/gaurav/avnc/ui/vnc/FrameView.kt:24]:
```suggestion
```

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -57,6 +57,7 @@
     <string name="msg_server_profile_added">Server added.</string>
     <string name="msg_server_profile_deleted">Server deleted.</string>
     <string name="msg_pip_not_supported">Picture-in-Picture is not supported on this device</string>
+    <string name="msg_ptr_hiding_not_supported">Hiding cursor not supported on this device</string>
```
Comment (gujjwal00) [inline app/src/main/res/values/strings.xml:60]:
```suggestion
    <string name="msg_ptr_hiding_not_supported">Hiding cursor is not supported on this device</string>
```

File: app/src/main/java/com/gaurav/avnc/ui/vnc/FrameView.kt (hunk 9 above)
Comment (gujjwal00) [reply app/src/main/java/com/gaurav/avnc/ui/vnc/FrameView.kt:24]: These imports are not needed anymore.

**Comment by gujjwal00** [comment]
Thank you.
